libfwupd/fwupd-client.c

```diff
@@ -121,13 +123,116 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(FwupdCurlHelper, fwupd_client_curl_helper_free)
 #endif
 
+typedef struct {
+	FwupdClient	*self;
+	gchar		*property_name;
+	guint		 signal_id;
+	FwupdDevice	*device;
+} FwupdClientContextHelper;
+
+static void
+fwupd_client_context_helper_free (FwupdClientContextHelper *helper)
+{
+	if (helper->device != NULL)
+		g_object_unref (helper->device);
+	g_object_unref (helper->self);
+	g_free (helper->property_name);
+	g_free (helper);
+}
+
+static gboolean
+fwupd_client_context_idle_cb (gpointer user_data)
+{
+	FwupdClientContextHelper *helper = (FwupdClientContextHelper *) user_data;
+	FwupdClient *self = FWUPD_CLIENT (helper->self);
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = NULL;
+
+	/* property */
+	if (helper->property_name != NULL)
+		g_object_notify (G_OBJECT (self), helper->property_name);
+
+	/* legacy signal name */
+	if (g_strcmp0 (helper->property_name, "status") == 0)
+		g_signal_emit (self, signals[SIGNAL_STATUS_CHANGED], 0, priv->status);
+
+	/* device signal */
+	if (helper->property_name != NULL)
+		g_signal_emit (self, signals[helper->signal_id], 0, helper->device);
+
+	/* all done */
+	locker = g_mutex_locker_new (&priv->idle_id_mutex);
+	priv->idle_id = 0;
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fwupd_client_context_helper (FwupdClient *self, FwupdClientContextHelper *helper)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_id_mutex);
+	g_autoptr(GSource) source = g_idle_source_new ();
+
+	/* run in the correct GMainContext and thread */
+	g_source_set_callback (source,
+			       fwupd_client_context_idle_cb,
+			       helper,
+			       (GDestroyNotify) fwupd_client_context_helper_free);
+	if (priv->idle_id != 0)
+		g_source_remove (priv->idle_id);
```

Comment:
This will cause the previously-pending signal emission to never happen, as the state from the `helper` object attached to the source being removed here is not stored anywhere else.

Reply:
That was by design; is that right? My logic was that if nothing was "servicing" the GMainContext then it's pointless just queing them all up for when the method returns.

Reply:
I don’t think it’s right. It’s quite plausible for the `FwupdClient` thread to make two `notify` emissions in quick succession, fast enough that the main thread might not be able to process the first one before it’s removed and replaced with the second one.

If you want to do this properly, I’d suggest keeping a `GPtrArray` of the `GSource`s which are pending. Just store the pointers to the `GSource`s, don’t bother with the source IDs (they’re equivalent).

Reply:
> keeping a GPtrArray of the GSources which are pending

How would I remove the GSource after the idle source has run? Surely a better way would be to store the FwupdClientContextHelper's in an array as then you have something in fwupd_client_context_idle_cb() to remove. i.e. There's only one idle_id like now, but when fwupd_client_context_idle_cb is called we might emit multiple signals...

Reply:
Sure, that would work better

Reply:
> Sure, that would work better

**Narrator:** It did!